### ISSUE TITLE
Intermittent issue "Invalid token header. No credentials provided."

### ISSUE THREAD

**Issue by daesu** [PhilipGarnero/django-rest-framework-social-oauth2#89]
Hi all, 

I followed the instructions on github and got up and running in no time. Everything was working well but for no apparent reason it seems to intermittently fail to detect the 'Authentication' Header and gives the following error; 

_"detail": "Invalid token header. No credentials provided."_

I am exchanging a facebook token for a DOT one as per the instructions for social signin and I successfully get a token back. But subsequent calls to my endpoints with the DOT token is failing. 

It's quite strange because it really does seem intermittent and seems to work fine one moment and then every endpoint is rejected the next.

Any suggestions would be much appreciated.

Env: 
Python 3.4.3
Django==1.10.5
django-oauth-toolkit==0.12.0
djangorestframework==3.5.3
django-rest-framework-social-oauth2==1.0.5

#Settings Snippet
```
INSTALLED_APPS = [
    <snip>

    'oauth2_provider',
    'social_django',
    'rest_framework_social_oauth2',
]

OAUTH2_PROVIDER = {
    'OAUTH2_BACKEND_CLASS': 'oauth2_provider.oauth2_backends.JSONOAuthLibCore'
}

#### Move to Config
SOCIAL_AUTH_USERNAME_IS_FULL_EMAIL = True

## Facebook configuration
SOCIAL_AUTH_FACEBOOK_KEY = 'snip'
SOCIAL_AUTH_FACEBOOK_SECRET = 'snip'

# Define SOCIAL_AUTH_FACEBOOK_SCOPE to get extra permissions from facebook. Email is not sent by default, to get it, you must request the email permission:
SOCIAL_AUTH_FACEBOOK_SCOPE = ['email']
SOCIAL_AUTH_FACEBOOK_PROFILE_EXTRA_PARAMS = {
    'fields': 'id, name, email'
}

REST_FRAMEWORK = {
    'DEFAULT_PERMISSION_CLASSES': [
        'rest_framework.permissions.IsAuthenticated',
    ],
    'PAGE_SIZE': 15,
    'DATETIME_FORMAT': "%d-%b-%y",
    'COERCE_DECIMAL_TO_STRING': False,
    'DEFAULT_AUTHENTICATION_CLASSES': (
        # OAuth
        'oauth2_provider.ext.rest_framework.OAuth2Authentication',
        'rest_framework_social_oauth2.authentication.SocialAuthentication',
    )
}

AUTHENTICATION_BACKENDS = (
    # Facebook OAuth2
    'social_core.backends.facebook.FacebookAppOAuth2',
    'social_core.backends.facebook.FacebookOAuth2',

    # #Social Auth
    'rest_framework_social_oauth2.backends.DjangoOAuth2',
    # Django
    'django.contrib.auth.backends.ModelBackend',
)

MIDDLEWARE = [
    'corsheaders.middleware.CorsMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.security.SecurityMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
    'django.middleware.locale.LocaleMiddleware',
    'oauth2_provider.middleware.OAuth2TokenMiddleware',
]

......

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': ['templates'],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.template.context_processors.i18n',
                'django.contrib.messages.context_processors.messages',
                # Social Auth
                'social_django.context_processors.backends',
                'social_django.context_processors.login_redirect',
            ],
        },
    },
]
```

**Comment by PhilipGarnero** [comment]
Could you show me your request as well ?

**Comment by daesu** [comment]
Hi Philip. 

So for example (using POSTman)

1. I make a call to /auth/convert-token with the following params; 

Headers: Content-Type: application/json

{
 "grant_type": "convert_token",  
  "client_id": "<snip>",
  "client_secret": "<snip>",
  "backend": "facebook",
 "token": "<facebook token>"
}

And get a response like so;

{
  "token_type": "Bearer",
  "access_token": "d26W2SNzrVpEotk9UzrZe1i01NdjVE",
  "refresh_token": "LtwC3Oytcst9dikxQwZqBDKHb1ZY2S",
  "scope": "read write",
  "expires_in": 36000
}

2. When I make a call to auth/token/ with the following params;

Headers: 
Content-Type: application/json
Authorization: Bearer d26W2SNzrVpEotk9UzrZe1i01NdjVE

{
 "grant_type": "refresh_token",  
  "client_id": "<snip>",
  "client_secret": "<snip>",
  "backend": "facebook",
 "refresh_token": "LtwC3Oytcst9dikxQwZqBDKHb1ZY2S"
}

IT gives the response;


It gives the response; 

{
  "detail": "Invalid token header. No credentials provided."
}

**Comment by PhilipGarnero** [comment]
Your refresh_token call is wrong.
You don't need to specify the backend or to put any headers.
Remove the backend and the headers and you should be fine.

Close this issue if that fixes your problem.

**Comment by daesu** [comment]
Hi Philip. 

I removed the headers and the raw POSTman code looks like;

curl -X POST   http://serveraddress.com/auth/token/ -d 'grant_type=refresh_token&client_id=snip&client_secret=snip&refresh_token=YScgciVmm24LWdgykB09f3gIvx7qsj'

But I am still getting; {"error":"invalid_grant"}

Some further information;

The Application settings are;
Client type: Confidential 
Authorization grant type: Resource owner password-based

**Comment by PhilipGarnero** [comment]
Hum, not sure why this doesn't work. It should.
Could you try to post it as json ?

**Comment by daesu** [comment]
Hi Philip. 

When I post as JSON like so;

curl -X POST   http://serveraddress.com/auth/token/  -d '{
 "grant_type": "refresh_token",  
  "client_id": "snip",
  "client_secret": "snip",
 "refresh_token": "LtwC3Oytcst9dikxQwZqBDKHb1ZY2S"
}'

It gives me; 

{"error":"unsupported_grant_type"

**Comment by PhilipGarnero** [comment]
I don't see why it wouldn't work.
Try to [debug](https://docs.python.org/2/library/pdb.html#debugger-commands) by putting `import pdb; pdb.set_trace()` [here](https://github.com/PhilipGarnero/django-rest-framework-social-oauth2/blob/master/rest_framework_social_oauth2/views.py#L41) inside `env/lib/python2.7/site-packages/rest_framework_social_oauth2/views.py` in your virtualenv  and check if the request contains everything you submitted. If it does, continue debugging by stepping through functions.
Post here, anything you've found.
I can't really help you more than that sorry.

**Comment by daesu** [comment]
Hi Philip, 

I just removed the old Application and re-created it according to the instructions and it now appears to be working. I have done this on both local and a remote Heroku instance. 

I think the issue may have been I hadn't specified a superuser account for the Application when I first created it but subsequently pointed it to the superuser. When I made the new application I set the superuser on creation. 

Thank you for all your help regardless and thank you for this really helpful lib.
